File: spring-data-requery-java/src/test/java/com/coupang/springframework/data/requery/repository/support/RequeryRepositoryFactoryBeanTest.java
package com.coupang.springframework.data.requery.repository.support;

import com.coupang.springframework.data.requery.core.RequeryOperations;
import com.coupang.springframework.data.requery.domain.basic.AbstractBasicUser;
import com.coupang.springframework.data.requery.repository.RequeryRepository;
import io.requery.meta.EntityModel;
import org.jetbrains.annotations.NotNull;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnitRunner;
import org.springframework.beans.factory.BeanFactory;
import org.springframework.beans.factory.ListableBeanFactory;
import org.springframework.dao.support.PersistenceExceptionTranslator;
import org.springframework.data.repository.Repository;
import org.springframework.data.repository.core.EntityInformation;
import org.springframework.data.repository.core.RepositoryInformation;
import org.springframework.data.repository.core.RepositoryMetadata;
import org.springframework.data.repository.core.support.RepositoryComposition;
import org.springframework.data.repository.core.support.RepositoryFactorySupport;

import java.util.HashMap;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

/**
 * RequeryRepositoryFactoryBeanTest
 *
 * @author debop@coupang.com
 * @since 18. 6. 12
 */
@RunWith(MockitoJUnitRunner.Silent.class)
public class RequeryRepositoryFactoryBeanTest {

    RequeryRepositoryFactoryBean<SimpleSampleRepository, AbstractBasicUser, Long> factoryBean;

    @Mock RequeryOperations operations;
    StubRepositoryFactorySupport factory;
    @Mock ListableBeanFactory beanFactory;
    @Mock PersistenceExceptionTranslator translator;
    @Mock Repository<?, ?> repository;
    @Mock EntityModel entityModel;

    @Before
    public void setup() {
        Map<String, PersistenceExceptionTranslator> beans = new HashMap<>();
        beans.put("foo", translator);

        when(beanFactory.getBeansOfType(eq(PersistenceExceptionTranslator.class), anyBoolean(), anyBoolean()))
            .thenReturn(beans);

        when(operations.getEntityModel()).thenReturn(entityModel);

        factory = Mockito.spy(new StubRepositoryFactorySupport(repository));

        // Setup standard factory configuration
        factoryBean = new DummyRequeryRepositoryFactoryBean<>(SimpleSampleRepository.class);
        factoryBean.setOperations(operations);
    }

    @Test
    public void setsBasicInstanceCorrectly() {

        factoryBean.setBeanFactory(beanFactory);
        factoryBean.afterPropertiesSet();

        assertThat(factoryBean.getObject()).isNotNull();
    }

    @Test(expected = IllegalArgumentException.class)
    public void requiresListableBeanFactory() {

        factoryBean.setBeanFactory(mock(BeanFactory.class));

    }

    @Test(expected = IllegalArgumentException.class)
    public void preventsNullRepositoryInterface() {
        new RequeryRepositoryFactoryBean<Repository<Object, Long>, Object, Long>(null);
    }


    private class DummyRequeryRepositoryFactoryBean<T extends RequeryRepository<S, ID>, S, ID> extends RequeryRepositoryFactoryBean<T, S, ID> {

        public DummyRequeryRepositoryFactoryBean(@NotNull Class<? extends T> repositoryInterface) {
            super(repositoryInterface);
        }

        @Override
        protected RepositoryFactorySupport doCreateRepositoryFactory() {
            return factory;
        }
    }


    private interface SimpleSampleRepository extends RequeryRepository<AbstractBasicUser, Long> {

    }

    private static class StubRepositoryFactorySupport extends RepositoryFactorySupport {

        private final Repository<?, ?> repository;

        private StubRepositoryFactorySupport(Repository<?, ?> repository) {
            this.repository = repository;
        }

        @SuppressWarnings("unchecked")
        @Override
        public <T> T getRepository(Class<T> repositoryInterface, RepositoryComposition.RepositoryFragments fragments) {
            return (T) repository;
        }

        @Override
        public <T, ID> EntityInformation<T, ID> getEntityInformation(Class<T> domainClass) {
            return null;
        }

        @Override
        protected Object getTargetRepository(RepositoryInformation metadata) {
            return null;
        }

        @Override
        protected Class<?> getRepositoryBaseClass(RepositoryMetadata metadata) {
            return null;
        }
    }
}

File: spring-data-requery-kotlin/src/test/kotlin/org/springframework/data/requery/kotlin/repository/support/RequeryRepositoryFactoryBeanTest.kt
package org.springframework.data.requery.kotlin.repository.support

import com.nhaarman.mockito_kotlin.*
import io.requery.meta.EntityModel
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.junit.MockitoJUnitRunner
import org.springframework.beans.factory.ListableBeanFactory
import org.springframework.dao.support.PersistenceExceptionTranslator
import org.springframework.data.repository.Repository
import org.springframework.data.repository.core.EntityInformation
import org.springframework.data.repository.core.RepositoryInformation
import org.springframework.data.repository.core.RepositoryMetadata
import org.springframework.data.repository.core.support.RepositoryComposition
import org.springframework.data.repository.core.support.RepositoryFactorySupport
import org.springframework.data.requery.kotlin.core.RequeryOperations
import org.springframework.data.requery.kotlin.domain.sample.User
import org.springframework.data.requery.kotlin.repository.RequeryRepository
import org.springframework.data.requery.kotlin.uninitialized

@RunWith(MockitoJUnitRunner.Silent::class)
class RequeryRepositoryFactoryBeanTest {

    private lateinit var factoryBean: RequeryRepositoryFactoryBean<SimpleSampleRepository, User, Int>
    private lateinit var factory: StubRepositoryFactorySupport

    val operations = mock<RequeryOperations>()
    val beanFactory = mock<ListableBeanFactory>()
    val translator = mock<PersistenceExceptionTranslator>()
    val repository = mock<Repository<out Any, out Any>>()
    val entityModel = mock<EntityModel>()

    @Before
    fun setup() {
        val beans = hashMapOf("foo" to translator)


        whenever(beanFactory.getBeansOfType(eq(PersistenceExceptionTranslator::class.java), any(), any()))
            .doReturn(beans)

        whenever(operations.entityModel).doReturn(entityModel)

        factory = spy(StubRepositoryFactorySupport(repository))

        // Setup standard factory configuration
        factoryBean = DummyRequeryRepositoryFactoryBean(SimpleSampleRepository::class.java)
        factoryBean.setOperations(operations)
    }

    @Test
    fun `set basic instance correctly`() {

        factoryBean.setBeanFactory(beanFactory)
        factoryBean.afterPropertiesSet()

        assertThat(factoryBean.`object`).isNotNull
    }

    @Test
    fun `requires listable bean factory`() {
        assertThatThrownBy {
            factoryBean.setBeanFactory(mock())
        }.isInstanceOf(IllegalArgumentException::class.java)
    }


    private inner class DummyRequeryRepositoryFactoryBean<T: RequeryRepository<E, ID>, E: Any, ID: Any>(repositoryInterface: Class<out T>)
        : RequeryRepositoryFactoryBean<T, E, ID>(repositoryInterface) {

        override fun doCreateRepositoryFactory(): RepositoryFactorySupport {
            return factory
        }
    }

    private interface SimpleSampleRepository: RequeryRepository<User, Int>

    @Suppress("UNCHECKED_CAST")
    private class StubRepositoryFactorySupport constructor(private val repository: Repository<*, *>): RepositoryFactorySupport() {

        override fun <T> getRepository(repositoryInterface: Class<T>, fragments: RepositoryComposition.RepositoryFragments): T {
            return repository as T
        }

        override fun <T, ID> getEntityInformation(domainClass: Class<T>): EntityInformation<T, ID> = uninitialized()

        override fun getTargetRepository(metadata: RepositoryInformation): Any = uninitialized()

        override fun getRepositoryBaseClass(metadata: RepositoryMetadata): Class<*> = uninitialized()
    }
}